File: sanitizers/tsan/cce/fortran/saxpy_openmp.f90
! This program performs a SAXPY operation using OpenMP
program saxpy_openmp
  use omp_lib
  implicit none

  integer, parameter :: N = 1000000
  real(kind=4) :: alpha = 2.0
  real(kind=4), dimension(N) :: X, Y
  integer :: i

  ! Initialize the arrays
  X = 1.0
  Y = 2.0

  ! Perform the SAXPY operation in parallel
  !$omp parallel do private(i) shared(X, Y, alpha)
  do i = 1, N
    Y(i) = alpha * X(i) + Y(i)
  end do
  !$omp end parallel do

  ! Print the result of the operation
  print *, 'The first element of Y after SAXPY is: ', Y(1)

end program saxpy_openmp

File: sanitizers/tsan/cce/cpp/saxpy_openmp.cpp
#include <omp.h>
#include <vector>
#include <iostream>

int main() {
    const int N = 1000000;
    float alpha = 2.0f;
    std::vector<float> X(N, 1.0f); // Initialize X with 1.0
    std::vector<float> Y(N, 2.0f); // Initialize Y with 2.0

    #pragma omp parallel for
    for(int i = 0; i < N; ++i) {
        Y[i] = alpha * X[i] + Y[i];
    }

    std::cout << "The first element of Y after SAXPY is: " << Y[0] << std::endl;

    return 0;
}
